src/module_wrap.cc

```diff
@@ -48,11 +48,11 @@ using v8::Undefined;
 using v8::Value;
 
 static const char* const EXTENSIONS[] = {
-  ".mjs",
-  ".cjs",
   ".js",
   ".json",
-  ".node"
+  ".node",
+  ".mjs",
```

Comment:
I don't think we check these in CJS. There is some risk that this leads to two valid resolutions, especially when combined with require hooks:

* `foo.jsx` or `foo.ts`: Found by CJS.
* `foo.mjs` or `foo.cjs`: Found by ESM.

Should we remove them? I think I'm fine keeping them since require hooks aren't "native".

Reply:
Don't think we should remove them personally, but maybe I'm not entirely understanding the case here

Reply:
> Should we remove them?

This brings up a good point about whether we should expect the resolution outcomes for `--es-module-specifier-resolution=node` and [`require.resolve()`](https://nodejs.org/api/modules.html#modules_require_resolve_request_options) to be identical.

The `rollup-plugin-node-resolve` package [refers to the current order of extension resolution](https://github.com/rollup/rollup-plugin-node-resolve/blob/master/src/index.js#L12-L14) as `node` (it's in the name).

Maybe a new mode of `legacy` (`--es-module-specifier-resolution=legacy`) should be introduced to function exactly like `require.resolve()`. The ESM package [refers to it as legacy](https://github.com/standard-things/esm/blob/950e74186acf7e2e733168b327f8d9747ad13427/src/module/esm/resolve-filename.js#L295) internally.

Reply:
Erm, couldn't we choose to pull the extension list and order from `require.extensions` directly, for compat? Even if it's the pre-user-code-modified version (therefore didn't support custom extensions), it'd at least guarantee they're in sync.

Reply:
What will be the consequences of removing?

Reply:
So i've gone ahead and left `.mjs` extension and removed the `.cjs` extension. Thought process here is that the `cjs` loader doesn't recognize that extension, so lets keep things consistent. They might be a bit weird... but at least they are primarily the same.

The CJS loader will currently fail on requiring a .mjs extension, not pass it through... so I think the behavior is equivalent.

This will pass through anything that doesn't match to the CJS loader, which in turn will get anything from require.extensions.

I believe this covers all the concerns raised in this thread

Reply:
I think removing the `.cjs` extension is a good start as I personally believe that the `node` resolution mode should function exactly as described in the [**All Together...** algorithm](https://nodejs.org/api/modules.html#modules_all_together).

In defense of modifying the behavior of this flag, the blog post that announced this flag made clear that its behavior would change once the ES module context went live unflagged, so removing or changing extension precedence hopefully won't come as too much of a shock to most.

As long as the `--loader` flag will continue to be available, I don't think anyone can justify being too upset about changes like this since the option of supplying one's own resolution algorithm is still on the table.